# Customer Churn Prediction

This project predicts whether a customer will churn or not using Machine Learning.

Internship: CodSoft Machine Learning Internship

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

print("Libraries imported successfully!")

Libraries imported successfully!


Unzip ZIP file

In [5]:
!unzip "archive (4).zip"

Archive:  archive (4).zip
  inflating: Churn_Modelling.csv     


Load dataset

In [6]:
import pandas as pd

df = pd.read_csv("Churn_Modelling.csv")

df.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


Check dataset

In [7]:
df.shape

(10000, 14)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


In [9]:
df["Exited"].value_counts()

,count
Exited,
0,7963
1,2037


Remove Unnecessary Columns

In [10]:
df = df.drop(["RowNumber", "CustomerId", "Surname"], axis=1)

df.head()

,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


Encode Gender

In [12]:
from sklearn.preprocessing import LabelEncoder

le_gender = LabelEncoder()
df["Gender"] = le_gender.fit_transform(df["Gender"])

df.head()

,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,619,France,0,42,2,0.00,1,1,1,101348.88,1
1,608,Spain,0,41,1,83807.86,1,0,1,112542.58,0
2,502,France,0,42,8,159660.80,3,1,0,113931.57,1
3,699,France,0,39,1,0.00,2,0,0,93826.63,0
4,850,Spain,0,43,2,125510.82,1,1,1,79084.10,0


Encode Geography

In [13]:
df = pd.get_dummies(df, columns=["Geography"], drop_first=True)

df.head()

,CreditScore,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,Geography_Germany,Geography_Spain
0,619,0,42,2,0.00,1,1,1,101348.88,1,False,False
1,608,0,41,1,83807.86,1,0,1,112542.58,0,False,True
2,502,0,42,8,159660.80,3,1,0,113931.57,1,False,False
3,699,0,39,1,0.00,2,0,0,93826.63,0,False,False
4,850,0,43,2,125510.82,1,1,1,79084.10,0,False,True


Separate X and y

In [14]:
X = df.drop("Exited", axis=1)
y = df["Exited"]

print(X.shape)
print(y.shape)

(10000, 11)
(10000,)


Train-test split

In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print(X_train.shape)
print(X_test.shape)

(8000, 11)
(2000, 11)


Feature scaling

In [16]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print("Scaling completed successfully!")

Scaling completed successfully!


In [ ]:
Training Models

Logistic Regression

In [17]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

lr_model = LogisticRegression()

lr_model.fit(X_train, y_train)

lr_pred = lr_model.predict(X_test)

lr_accuracy = accuracy_score(y_test, lr_pred)

print("Logistic Regression Accuracy:", lr_accuracy)

Logistic Regression Accuracy: 0.811


Random Forest

In [18]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(random_state=42)

rf_model.fit(X_train, y_train)

rf_pred = rf_model.predict(X_test)

rf_accuracy = accuracy_score(y_test, rf_pred)

print("Random Forest Accuracy:", rf_accuracy)

Random Forest Accuracy: 0.8665


Gradient Boosting

In [19]:
from sklearn.ensemble import GradientBoostingClassifier

gb_model = GradientBoostingClassifier(random_state=42)

gb_model.fit(X_train, y_train)

gb_pred = gb_model.predict(X_test)

gb_accuracy = accuracy_score(y_test, gb_pred)

print("Gradient Boosting Accuracy:", gb_accuracy)

Gradient Boosting Accuracy: 0.8675


Compare

In [20]:
print("Logistic Regression Accuracy:", lr_accuracy)
print("Random Forest Accuracy:", rf_accuracy)
print("Gradient Boosting Accuracy:", gb_accuracy)

Logistic Regression Accuracy: 0.811
Random Forest Accuracy: 0.8665
Gradient Boosting Accuracy: 0.8675


Choose best model

In [21]:
model_accuracies = {
    "Logistic Regression": lr_accuracy,
    "Random Forest": rf_accuracy,
    "Gradient Boosting": gb_accuracy
}

best_model_name = max(model_accuracies, key=model_accuracies.get)
best_accuracy = model_accuracies[best_model_name]

print("Best Model:", best_model_name)
print("Best Accuracy:", best_accuracy)

Best Model: Gradient Boosting
Best Accuracy: 0.8675


Save the Best Model

In [22]:
import joblib

joblib.dump(gb_model, "customer_churn_model.pkl")

print("Model Saved Successfully!")

Model Saved Successfully!


Download

In [23]:
from google.colab import files

files.download("customer_churn_model.pkl")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Test on Sample Customer

In [25]:
import pandas as pd

sample_customer = pd.DataFrame([[
    600,
    1,
    40,
    5,
    50000,
    2,
    1,
    1,
    70000,
    0,
    1
]], columns=X.columns)

sample_customer = scaler.transform(sample_customer)

prediction = gb_model.predict(sample_customer)

if prediction[0] == 1:
    print("Customer is likely to Churn")
else:
    print("Customer is likely to Stay")

Customer is likely to Stay


Test_2

In [29]:
sample_customer = pd.DataFrame([[
    450,      # CreditScore
    0,        # Female
    55,       # Age
    1,        # Tenure
    150000,   # Balance
    1,        # NumOfProducts
    0,        # HasCrCard
    0,        # IsActiveMember
    30000,    # EstimatedSalary
    1,        # Geography_Germany
    0         # Geography_Spain
]], columns=X.columns)
sample_customer = scaler.transform(sample_customer)

prediction = gb_model.predict(sample_customer)

if prediction[0] == 1:
    print("Customer is likely to Churn")
else:
    print("Customer is likely to Stay")

Customer is likely to Churn


Test_3

In [30]:
sample_customer = pd.DataFrame([[
    800,      # CreditScore
    1,        # Male
    35,       # Age
    8,        # Tenure
    20000,    # Balance
    2,        # NumOfProducts
    1,        # HasCrCard
    1,        # IsActiveMember
    120000,   # EstimatedSalary
    0,        # Geography_Germany
    1         # Geography_Spain
]], columns=X.columns)
sample_customer = scaler.transform(sample_customer)

prediction = gb_model.predict(sample_customer)

if prediction[0] == 1:
    print("Customer is likely to Churn")
else:
    print("Customer is likely to Stay")

Customer is likely to Stay


Conclusion

In this project, a Customer Churn Prediction model was developed using Machine Learning techniques. The dataset was preprocessed by handling categorical features, encoding data, and applying feature scaling.

Three machine learning algorithms were trained and evaluated:

• Logistic Regression – 81.10%
• Random Forest – 86.65%
• Gradient Boosting – 86.75%

Among all the models, Gradient Boosting achieved the highest accuracy of 86.75% and was selected as the final model.

The model can help businesses identify customers who are likely to leave their services, allowing them to take proactive measures to improve customer retention and reduce churn.